src/Microsoft.Diagnostics.Runtime/src/DataReaders/MacOS/MacOSProcessDataReader.cs

```diff
@@ -221,10 +225,62 @@ private unsafe T Read<T>(ref ulong address)
             return result;
         }
 
-        public bool GetThreadContext(uint threadID, uint contextFlags, Span<byte> context)
+        public unsafe bool GetThreadContext(uint threadID, uint contextFlags, Span<byte> context)
         {
-            // TODO
-            return false;
+            LoadThreads();
+            if (!_threadActs.TryGetValue(threadID, out var threadAct))
+                return false;
+
+            int regSize = sizeof(x86_thread_state64_t);
+
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(regSize);
+            try
+            {
+                fixed (byte* data = buffer)
+                {
+                    int stateCount = Native.x86_THREAD_STATE64_COUNT;
+                    int kr = Native.thread_get_state(threadAct, Native.x86_THREAD_STATE64, new IntPtr(data), ref stateCount);
+                    if (kr != 0)
+                        throw new ClrDiagnosticsException($"thread_get_state failed with status code 0x{kr:x}");
+                }
+
+                Unsafe.As<byte, x86_thread_state64_t>(ref MemoryMarshal.GetReference(buffer.AsSpan())).CopyContext(context);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+
+            return true;
+        }
+
+        private unsafe void LoadThreads()
```

Comment:
Similarly, it would be good to not throw in this function if threads cannot be loaded.

Reply:
Hi! Thanks for your review. I agree with the another one, but in this case I'm not sure if this is what we want. It behaves similarly to private `LinuxLiveDataReader.LoadThreads` and `MacOSProcessDataReader.LoadMemoryRegions`.

Do you suggest to get Native.thread_info() for all threads, even if some of them fail?

Reply:
Good point.  The behavior should be consistent across all platforms, but it's not today...  I need to go back and take a look and make it all consistent.

In general, we want functions which return `bool` to return false on failure, and push that upwards in the stack.  So LoadThreads both here and in Linux should probably have a `bool` return value instead of throwing.

The intent is that even if we cannot walk threads or walk callstacks, a tool should still be able to walk the heap without hitting exceptions.  I'll flag an issue to take a look at this more broadly.  We won't block this checkin on that though since, as you point out, it's just not consistent today.

FYI, I am out sick today.  I will take a closer look at the other Pull Requests on monday.